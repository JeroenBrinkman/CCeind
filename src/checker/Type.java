package checker;

import iloc.eval.Machine;

/** TempName data type. */
abstract public class Type {
	/** The singleton instance of the {@link Bool} type. */
	public static final Type BOOL = new Bool();
	/** The singleton instance of the {@link Int} type. */
	public static final Type INT = new Int();
	/** The singleton instance of the {@link Char} type. */
	public static final Type CHAR = new Char();
	/** The singleton instance of the {@link String} type. */
	public static final Type STRING = new String();
	/** The singleton instance of the {@link Void} type. */
	public static final Type VOID = new Void();
	
	private final TypeKind kind;

	/** Constructor for subclasses. */
	protected Type(TypeKind kind) {
		this.kind = kind;
	}

	/** Returns the kind of this type. */
	public TypeKind getKind() {
		return this.kind;
	}

	/** returns the size (in bytes) of a value of this type. */
	abstract public int size();

	/** Representation of the Boolean type. */
	static public class Bool extends Type {
		private Bool() {
			super(TypeKind.BOOL);
		}

		@Override
		public int size() {
			return Machine.INT_SIZE;
		}

		@Override
		public java.lang.String toString() {
			return "Boolean";
		}
	}

	/** Representation of the Integer type. */
	static public class Int extends Type {
		private Int() {
			super(TypeKind.INT);
		}

		@Override
		public int size() {
			return Machine.INT_SIZE;
		}

		@Override
		public java.lang.String toString() {
			return "Integer";
		}
	}
	
	/** Representation of the Char type. */
	static public class Char extends Type{

		private Char(){
			super(TypeKind.CHAR);
		}
		
		@Override
		public int size() {
			return Machine.DEFAULT_CHAR_SIZE;
		}

		@Override
		public java.lang.String toString() {
			return "Character";
		}
		
	}
	
	/** Representation of the String type. */
	static public class String extends Type{
		private int length = 0;
		private String(){
			super(TypeKind.STRING);
		}
		
		@Override
		public int size() {
			return length*Machine.DEFAULT_CHAR_SIZE;
		}

		@Override
		public java.lang.String toString() {
			return "String";
		}
		
		public void setLength(int length){
			this.length = length;
			
		}
		
	}
	
	/** Representation of the Void type. */
	static public class Void extends Type{

		private Void(){
			super(TypeKind.VOID);
		}
		
		@Override
		public int size() {
			return 0;
		}

		@Override
		public java.lang.String toString() {
			return "Void";
		}
		
	}
}
