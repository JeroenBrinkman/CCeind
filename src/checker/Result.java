package checker;

import org.antlr.v4.runtime.ParserRuleContext;
import org.antlr.v4.runtime.tree.ParseTree;
import org.antlr.v4.runtime.tree.ParseTreeProperty;

/** Class holding the results of the TempName checker. */
public class Result {
	/**
	 * Mapping from statements and expressions to the atomic subtree that is
	 * their entry in the control flow graph.
	 */
	private final ParseTreeProperty<ParserRuleContext> entries = new ParseTreeProperty<>();
	/** Mapping from expressions to types. */
	private final ParseTreeProperty<Type> types = new ParseTreeProperty<>();
	private final ParseTreeProperty<Type[]> readtypes = new ParseTreeProperty<>();

	/** Adds an association from parse tree node to the flow graph entry. */
	public void setEntry(ParseTree node, ParserRuleContext entry) {
		this.entries.put(node, entry);
	}

	/**
	 * Returns the flow graph entry associated with a given parse tree node.
	 */
	public ParserRuleContext getEntry(ParseTree node) {
		return this.entries.get(node);
	}

	/**
	 * Adds an association from a parse tree expression, type, or assignment
	 * target node to the corresponding (inferred) type.
	 */
	public void setType(ParseTree node, Type type) {
		this.types.put(node, type);
	}

	/** Returns the type associated with a given parse tree node. */
	public Type getType(ParseTree node) {
		return this.types.get(node);
	}

	public void setReadTypes(ParseTree node, Type[] tps) {
		this.readtypes.put(node, tps);
	}

	public Type[] getReadTypes(ParseTree node) {
		return this.readtypes.get(node);
	}
}
