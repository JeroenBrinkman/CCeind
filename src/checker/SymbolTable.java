package checker;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

public class SymbolTable{

	private final Deque<Map<String, Type>> stack = new ArrayDeque<Map<String,Type>>();
	private List<String> constants = new ArrayList<String>();
	
	public void openScope() {
		stack.push(new HashMap<String, Type>());
	}

	public void closeScope() {
		stack.pop();
	}

	public boolean add(String id, Type type) {
		if(contains(id) || isConst(id)){
			return false;
			
		}else{
			Map<String, Type> currentScope;
			currentScope = stack.getFirst();
			currentScope.put(id, type);
			return true;
		}
	}
	
	public void addConst(String id){
		constants.add(id);
	}

	public boolean contains(String id) {
		Map<String, Type> currentScope;
		for(Iterator<Map<String, Type>> itr = stack.iterator(); itr.hasNext();){
			currentScope = (Map<String, Type>) itr.next();
			if(currentScope.containsKey(id)){
				return true;
			}
		}
		return false;
	}
	
	public boolean isConst(String id){
		return constants.contains(id);
	}
	
	public Type getType(String id){
		Type type = Type.VOID;
		Map<String, Type> currentScope;
		for(Iterator<Map<String, Type>> itr = stack.descendingIterator(); itr.hasNext();){
			currentScope = (Map<String, Type>) itr.next();
			if(currentScope.containsKey(id)){
				type = currentScope.get(id);
			}
		}
		return type;
	}
	
}
