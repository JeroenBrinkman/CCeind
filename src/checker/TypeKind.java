package checker;

/** TempName data type kind. */
public enum TypeKind {
	/** Integer base type. */
	INT,
	/** Boolean base type. */
	BOOL,
	/** Character base Type*/
	CHAR,
	/** String base Type*/
	STRING,
	/** Void base Type*/
	VOID;
}
