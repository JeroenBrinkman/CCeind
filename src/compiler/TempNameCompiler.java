package compiler;


import java.io.File;
import java.io.FileReader;
import java.io.IOException;

import org.antlr.v4.runtime.ANTLRInputStream;
import org.antlr.v4.runtime.CharStream;
import org.antlr.v4.runtime.CommonTokenStream;
import org.antlr.v4.runtime.Lexer;
import org.antlr.v4.runtime.TokenStream;
import org.antlr.v4.runtime.tree.ParseTree;

import checker.ErrorListener;
import checker.ParseException;
import checker.Result;
import checker.TypeChecker;
import grammar.TempNameLexer;
import grammar.TempNameParser;

public class TempNameCompiler {
	/** The singleton instance of this class. */
	private final static TempNameCompiler instance = new TempNameCompiler();
	/** Debug flag. */
	private final static boolean SHOW = true;

	/** Returns the singleton instance of this class. */
	public static TempNameCompiler instance() {
		return instance;
	}

	/** Compiles and runs the program named in the argument. */
	public static void main(String[] args) {
		if (args.length != 1) {
			System.err.println("Usage: filename");
			return;
		}
		//try {
			System.out.println("--- Running " + args[0]);
			//Program prog = instance().compile(new File(args[0]));
			if (SHOW) {
				//System.out.println(prog.prettyPrint());
			}
			//Simulator sim = new Simulator(prog);
			//sim.run();
			System.out.println("--- Done with " + args[0]);
		/**} catch (ParseException exc) {
			exc.print();
		} catch (IOException exc) {
			exc.printStackTrace();
		}*/
	}

	/** The fixed checker of this compiler. */
	private final TypeChecker checker;
	/** The fixed generator of this compiler. */
	//private final Generator generator;

	private TempNameCompiler() {
		this.checker = new TypeChecker();
		//this.generator = new Generator();
	}

	/** Typechecks a given Simple Pascal string. */
	public Result check(String text) throws ParseException {
		return check(parse(text));
	}

	/** Typechecks a given Simple Pascal file. */
	public Result check(File file) throws ParseException, IOException {
		return check(parse(file));
	}

	/** Typechecks a given Simple Pascal parse tree. */
	public Result check(ParseTree tree) throws ParseException {
		return this.checker.check(tree);
	}

	/** Compiles a given Simple Pascal string into an ILOC program. 
	public Program compile(String text) throws ParseException {
		return compile(parse(text));
	}*/

	/** Compiles a given Simple Pascal file into an ILOC program. 
	public Program compile(File file) throws ParseException, IOException {
		return compile(parse(file));
	}*/

	/** Compiles a given Simple Pascal parse tree into an ILOC program. 
	public Program compile(ParseTree tree) throws ParseException {
		Result checkResult = this.checker.check(tree);
		
		return this.generator.generate(tree, checkResult);
	}*/

	/** Compiles a given Simple Pascal string into a parse tree. */
	public ParseTree parse(String text) throws ParseException {
		return parse(new ANTLRInputStream(text));
	}

	/** Compiles a given Simple Pascal string into a parse tree. */
	public ParseTree parse(File file) throws ParseException, IOException {
		return parse(new ANTLRInputStream(new FileReader(file)));
	}

	private ParseTree parse(CharStream chars) throws ParseException {
		ErrorListener listener = new ErrorListener();
		Lexer lexer = new TempNameLexer(chars);
		lexer.removeErrorListeners();
		lexer.addErrorListener(listener);
		TokenStream tokens = new CommonTokenStream(lexer);
		TempNameParser parser = new TempNameParser(tokens);
		parser.removeErrorListeners();
		parser.addErrorListener(listener);
		ParseTree result = parser.program();
		System.out.println(result.toString());
		listener.throwException();
		return result;
	}
}
