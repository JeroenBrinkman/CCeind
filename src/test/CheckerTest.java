package test;

import static org.junit.Assert.*;

import java.io.File;
import java.io.IOException;

import org.antlr.v4.runtime.tree.ParseTree;
import org.junit.Test;

import checker.ParseException;
import checker.Result;
import checker.Type;
import compiler.TempNameCompiler;

public class CheckerTest {
	private final static String BASE_DIR = "src/files";
	private final static String EXT = ".tmp";
	private final TempNameCompiler compiler = TempNameCompiler.instance();

	@Test
	public void test0() throws ParseException, IOException {
		ParseTree tree = parse("test0");
		Result result = check(tree);
		ParseTree assX = tree.getChild(1).getChild(3);
		System.out.println(assX.getText());
		ParseTree assXX = assX.getChild(2);
		System.out.println(assXX.getText());
		assertEquals(Type.INT, result.getType(assX));
		assertEquals(Type.INT, result.getType(assXX.getChild(0)));
		assertEquals(Type.INT, result.getType(assXX.getChild(2)));
	}
	
	@Test
	public void test1() throws ParseException, IOException {
		ParseTree tree = parse("test1");
		System.out.println(tree.toStringTree());
		Result result = check(tree);
		ParseTree a = tree.getChild(1).getChild(1);
		System.out.println(a.getText());
		assertEquals(Type.VOID, result.getType(a));
		assertEquals(Type.INT, result.getType(a.getChild(3)));
		ParseTree print = tree.getChild(1).getChild(7);
		assertEquals(Type.INT, result.getType(print.getChild(2))); 
		assertEquals(Type.INT, result.getType(print));
	}

	@Test
	public void test2() throws ParseException, IOException {
		ParseTree tree = parse("test2");
		System.out.println(tree.toStringTree());
		Result result = check(tree);
	}
	
	@Test
	public void test3() throws ParseException, IOException {
		ParseTree tree = parse("test3");
		System.out.println(tree.toStringTree());
		Result result = check(tree);
	}
	
	private ParseTree parse(String filename) throws IOException, ParseException {
		return this.compiler.parse(new File(BASE_DIR, filename + EXT));
	}

	private Result check(ParseTree tree) throws ParseException {

		return this.compiler.check(tree);
	}
}
