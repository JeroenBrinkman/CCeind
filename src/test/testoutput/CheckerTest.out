----Testing Basics----
([] start ([19] { ([79 19] ([28 79 19] integer) a) ; ([79 19] ([28 79 19] boolean) b) ; ([79 19] ([28 79 19] char) c) ; ([79 19] ([28 79 19] string) d) ; ([79 19] ([34 79 19] a) := ([36 79 19] 1337)) ; ([79 19] ([34 79 19] b) := ([36 79 19] true)) ; ([79 19] ([34 79 19] c) := ([36 79 19] 'g')) ; ([79 19] ([34 79 19] d) := ([36 79 19] "herp derp")) ; ([79 19] ([28 79 19] integer) w := ([31 79 19] 69)) ; ([79 19] ([28 79 19] boolean) x := ([31 79 19] false)) ; ([79 19] ([28 79 19] char) y := ([31 79 19] 'g')) ; ([87 19] ([28 87 19] string) z := ([31 87 19] "Hello World!")) ; }) <EOF>)
----Testing Error 0----
Errors: 
Line 7:6 - Expected type 'Integer' but found 'Boolean'
Line 8:6 - Expected type 'Boolean' but found 'Character'
Line 9:6 - Expected type 'Character' but found 'String'
Line 10:6 - Expected type 'String' but found 'Integer'
Line 13:14 - Expected type 'Integer' but found 'Boolean'
Line 14:14 - Expected type 'Boolean' but found 'Character'
Line 15:11 - Expected type 'Character' but found 'String'
Line 16:13 - Expected type 'String' but found 'Integer'
----Testing Error 1----
Errors: 
Line 2:18 - Expected type 'Integer' but found 'Character'
Line 3:18 - Expected type 'Integer' but found 'String'
Line 4:15 - Expected type 'Integer' but found 'String'
Line 5:18 - Expected type 'Boolean' but found 'Integer'
Line 6:14 - Expected type 'Boolean' but found 'Integer'
Line 6:20 - Expected type 'Boolean' but found 'Integer'
Line 7:14 - Expected type 'Boolean' but found 'Integer'
Line 7:19 - Expected type 'Boolean' but found 'Integer'
Line 8:14 - Expected type 'Integer' but found 'Character'
Line 8:20 - Expected type 'Integer' but found 'Character'
Line 8:14 - Expected type 'Integer' but found 'Boolean'
Line 9:19 - Expected type 'Integer' but found 'Character'
Line 9:14 - Expected type 'Integer' but found 'Boolean'
----Testing Error 2----
Errors: 
Line 8:1 - Illegal declaration, 'aa' already in scope.
Line 9:1 - Illegal declaration, 'ab' already in scope.
Line 10:1 - Illegal declaration, 'ac' already in scope.
Line 11:1 - Illegal declaration, 'ad' already in scope.
Line 14:17 - Expected type 'Boolean' but found 'Integer'
Line 15:14 - Expected type 'Integer' but found 'Boolean'
Line 16:28 - Expected type 'Character' but found 'Integer'
Line 19:7 - Expected type 'Boolean' but found 'Integer'
Line 20:7 - Expected type 'Boolean' but found 'Integer'
Line 25:14 - Expected type 'Boolean' but found 'Character'
String reads are not supported, at : read(ad)
Line 26:14 - Expected type 'Character' but found 'String'
Line 27:14 - Expected type 'String' but found 'Integer'
Line 28:14 - Expected type 'Integer' but found 'Void'
Line 29:14 - Expected type 'Integer' but found 'Void'
Line 32:14 - Expected type 'Integer' but found 'Boolean'
Line 33:14 - Expected type 'Boolean' but found 'Character'
Line 34:14 - Expected type 'Character' but found 'String'
Line 35:14 - Expected type 'String' but found 'Integer'
Line 36:14 - Expected type 'Integer' but found 'Void'
Line 37:14 - Expected type 'Integer' but found 'Void'
Line 40:14 - Expected type 'Integer' but found 'Boolean'
Line 41:14 - Expected type 'Boolean' but found 'Character'
Line 42:14 - Expected type 'Character' but found 'String'
Line 43:14 - Expected type 'String' but found 'Integer'
Line 44:20 - Expected type 'Integer' but found 'Boolean'
Line 45:26 - Expected type 'Integer' but found 'Boolean'
Line 48:14 - Expected type 'Integer' but found 'Boolean'
Line 49:14 - Expected type 'Boolean' but found 'Character'
Line 50:14 - Expected type 'Character' but found 'String'
Line 51:14 - Expected type 'String' but found 'Integer'
Line 52:14 - Expected type 'Integer' but found 'Boolean'
----Testing Error 3----
Errors: 
line 1:10 - mismatched input 'ba' expecting {<EOF>, AND, OR, '==', '>=', '>', '<=', '<', '-', '%', '<>', '+', '/', '*'}
----Testing Expressions----
([] start ([19] { ([79 19] ([28 79 19] integer) a := ([31 79 19] 0)) ; ([79 19] ([28 79 19] integer) b := ([31 79 19] 1)) ; ([79 19] ([28 79 19] integer) c := ([31 79 19] 2)) ; ([79 19] ([28 79 19] integer) d := ([31 79 19] 4)) ; ([79 19] ([28 79 19] boolean) aa := ([31 79 19] true)) ; ([79 19] ([28 79 19] char) ab := ([31 79 19] 'd')) ; ([79 19] ([28 79 19] string) ac := ([31 79 19] "gg")) ; ([79 19] ([28 79 19] integer) e := ([31 79 19] if ([39 31 79 19] ([4 39 31 79 19] d) ([108 39 31 79 19] >) ([109 39 31 79 19] b)) then ([41 31 79 19] 69) else ([43 31 79 19] 96))) ; ([79 19] ([28 79 19] boolean) f := ([31 79 19] if ([39 31 79 19] ([4 39 31 79 19] c) ([108 39 31 79 19] <) ([109 39 31 79 19] d)) then ([41 31 79 19] true) else ([43 31 79 19] false))) ; ([79 19] ([28 79 19] char) g := ([31 79 19] if ([39 31 79 19] ([4 39 31 79 19] b) ([108 39 31 79 19] >) ([109 39 31 79 19] a)) then ([41 31 79 19] 'b') else ([43 31 79 19] 'a'))) ; ([79 19] ([28 79 19] string) h := ([31 79 19] if ([39 31 79 19] ([4 39 31 79 19] b) ([108 39 31 79 19] >) ([109 39 31 79 19] c)) then ([41 31 79 19] "gg") else ([43 31 79 19] "wp"))) ; ([79 19] while ([47 79 19] ([4 47 79 19] a) ([108 47 79 19] >) ([109 47 79 19] b)) do ([49 79 19] ([4 49 79 19] a) ([104 49 79 19] +) ([105 49 79 19] b))) ; ([79 19] while ([47 79 19] ([4 47 79 19] a) ([108 47 79 19] <) ([109 47 79 19] b)) do ([49 79 19] ([4 49 79 19] a) ([104 49 79 19] +) ([105 49 79 19] b))) ; ([79 19] while ([47 79 19] false) do ([49 79 19] ([4 49 79 19] a) ([104 49 79 19] -) ([105 49 79 19] b))) ; ([79 19] ([28 79 19] integer) i := ([31 79 19] read ( a ))) ; ([79 19] ([28 79 19] boolean) ia := ([31 79 19] read ( aa ))) ; ([79 19] ([28 79 19] char) ib := ([31 79 19] read ( ab ))) ; ([79 19] ([28 79 19] integer) j := ([31 79 19] print ( ([64 31 79 19] 69) ))) ; ([79 19] ([28 79 19] integer) k := ([31 79 19] print ( ([64 31 79 19] a) ))) ; ([79 19] ([28 79 19] integer) l := ([31 79 19] print ( ([64 31 79 19] ([4 64 31 79 19] c) ([100 64 31 79 19] *) ([101 64 31 79 19] d)) ))) ; ([79 19] ([28 79 19] boolean) m := ([31 79 19] print ( ([64 31 79 19] ([4 64 31 79 19] true) ([112 64 31 79 19] or) ([113 64 31 79 19] false)) ))) ; ([79 19] ([28 79 19] char) n := ([31 79 19] print ( ([64 31 79 19] 'd') ))) ; ([79 19] ([28 79 19] string) o := ([31 79 19] print ( ([64 31 79 19] "1337") ))) ; ([79 19] ([28 79 19] integer) oa := ([31 79 19] print ( ([64 31 79 19] if ([39 64 31 79 19] ([4 39 64 31 79 19] a) ([108 39 64 31 79 19] >) ([109 39 64 31 79 19] b)) then ([41 64 31 79 19] 4) else ([43 64 31 79 19] 5)) ))) ; ([79 19] ([28 79 19] integer) p := ([31 79 19] ( ([75 31 79 19] 4) ))) ; ([79 19] ([28 79 19] integer) q := ([31 79 19] ( ([75 31 79 19] ([4 75 31 79 19] a) ([104 75 31 79 19] +) ([105 75 31 79 19] b)) ))) ; ([79 19] ([28 79 19] boolean) r := ([31 79 19] ( ([75 31 79 19] if ([39 75 31 79 19] ([4 39 75 31 79 19] a) ([108 39 75 31 79 19] >) ([109 39 75 31 79 19] b)) then ([41 75 31 79 19] true) else ([43 75 31 79 19] false)) ))) ; ([79 19] ([28 79 19] char) s := ([31 79 19] ( ([75 31 79 19] 'a') ))) ; ([79 19] ([28 79 19] string) t := ([31 79 19] ( ([75 31 79 19] "Hello World!") ))) ; ([79 19] ([28 79 19] integer) u := ([31 79 19] ( ([75 31 79 19] ([4 75 31 79 19] 1) ([104 75 31 79 19] +) ([105 75 31 79 19] ( ([75 105 75 31 79 19] ([4 75 105 75 31 79 19] c) ([100 75 105 75 31 79 19] *) ([101 75 105 75 31 79 19] d)) ))) ))) ; ([79 19] ([28 79 19] integer) v := ([31 79 19] { ([87 31 79 19] 69) ; })) ; ([79 19] ([28 79 19] integer) w := ([31 79 19] { ([79 31 79 19] 'd') ; ([79 31 79 19] "gg") ; ([79 31 79 19] true) ; ([87 31 79 19] 69) ; })) ; ([79 19] ([28 79 19] boolean) x := ([31 79 19] { ([87 31 79 19] true) ; })) ; ([79 19] ([28 79 19] char) y := ([31 79 19] { ([87 31 79 19] 'd') ; })) ; ([87 19] ([28 87 19] string) z := ([31 87 19] { ([87 31 87 19] "derp") ; })) ; }) <EOF>)
----Testing Operations----
([] start ([19] { ([79 19] ([28 79 19] integer) a := ([31 79 19] ([4 31 79 19] 8) ([100 31 79 19] *) ([101 31 79 19] 4))) ; ([79 19] ([28 79 19] integer) b := ([31 79 19] ([4 31 79 19] 8) ([100 31 79 19] /) ([101 31 79 19] 4))) ; ([79 19] ([28 79 19] integer) c := ([31 79 19] ([4 31 79 19] 8) ([100 31 79 19] %) ([101 31 79 19] 4))) ; ([79 19] ([28 79 19] integer) d := ([31 79 19] ([4 31 79 19] a) ([100 31 79 19] *) ([101 31 79 19] b))) ; ([79 19] ([28 79 19] integer) e := ([31 79 19] ([4 31 79 19] a) ([100 31 79 19] /) ([101 31 79 19] b))) ; ([79 19] ([28 79 19] integer) f := ([31 79 19] ([4 31 79 19] a) ([100 31 79 19] %) ([101 31 79 19] b))) ; ([79 19] ([28 79 19] integer) g := ([31 79 19] ([4 31 79 19] 8) ([104 31 79 19] +) ([105 31 79 19] 4))) ; ([79 19] ([28 79 19] integer) h := ([31 79 19] ([4 31 79 19] 8) ([104 31 79 19] -) ([105 31 79 19] 4))) ; ([79 19] ([28 79 19] integer) i := ([31 79 19] ([4 31 79 19] g) ([104 31 79 19] +) ([105 31 79 19] h))) ; ([79 19] ([28 79 19] integer) j := ([31 79 19] ([4 31 79 19] g) ([104 31 79 19] -) ([105 31 79 19] h))) ; ([79 19] ([28 79 19] integer) k := ([31 79 19] ([25 31 79 19] -) ([26 31 79 19] 4))) ; ([79 19] ([28 79 19] integer) l := ([31 79 19] ([25 31 79 19] -) ([26 31 79 19] b))) ; ([79 19] ([28 79 19] boolean) m := ([31 79 19] ([25 31 79 19] not) ([26 31 79 19] true))) ; ([79 19] ([28 79 19] boolean) n := ([31 79 19] ([25 31 79 19] not) ([26 31 79 19] m))) ; ([79 19] ([28 79 19] boolean) o := ([31 79 19] ([4 31 79 19] true) ([112 31 79 19] and) ([113 31 79 19] false))) ; ([79 19] ([28 79 19] boolean) p := ([31 79 19] ([4 31 79 19] true) ([112 31 79 19] or) ([113 31 79 19] false))) ; ([79 19] ([28 79 19] boolean) q := ([31 79 19] ([4 31 79 19] o) ([112 31 79 19] and) ([113 31 79 19] p))) ; ([79 19] ([28 79 19] boolean) r := ([31 79 19] ([4 31 79 19] o) ([112 31 79 19] or) ([113 31 79 19] p))) ; ([79 19] ([28 79 19] boolean) s := ([31 79 19] ([4 31 79 19] 6) ([108 31 79 19] <=) ([109 31 79 19] 9))) ; ([79 19] ([28 79 19] boolean) t := ([31 79 19] ([4 31 79 19] 6) ([108 31 79 19] <) ([109 31 79 19] 9))) ; ([79 19] ([28 79 19] boolean) u := ([31 79 19] ([4 31 79 19] 6) ([108 31 79 19] >=) ([109 31 79 19] 9))) ; ([79 19] ([28 79 19] boolean) v := ([31 79 19] ([4 31 79 19] 6) ([108 31 79 19] >) ([109 31 79 19] 9))) ; ([79 19] ([28 79 19] boolean) w := ([31 79 19] ([4 31 79 19] 6) ([108 31 79 19] ==) ([109 31 79 19] 9))) ; ([79 19] ([28 79 19] boolean) x := ([31 79 19] ([4 31 79 19] 6) ([108 31 79 19] <>) ([109 31 79 19] 9))) ; ([79 19] ([28 79 19] boolean) y := ([31 79 19] ([4 31 79 19] true) ([108 31 79 19] ==) ([109 31 79 19] false))) ; ([79 19] ([28 79 19] boolean) z := ([31 79 19] ([4 31 79 19] true) ([108 31 79 19] <>) ([109 31 79 19] false))) ; ([79 19] ([28 79 19] boolean) aa := ([31 79 19] ([4 31 79 19] 'g') ([108 31 79 19] ==) ([109 31 79 19] 'g'))) ; ([87 19] ([28 87 19] boolean) ab := ([31 87 19] ([4 31 87 19] 'g') ([108 31 87 19] <>) ([109 31 87 19] 'g'))) ; }) <EOF>)
